src/trainer.py

```diff
@@ -1115,14 +1137,67 @@ def _save_checkpoint(self, training_state, phase="pretrain", new_best_macro=Fals
                 "metric_state_{}_epoch_{}{}.th".format(phase, epoch, best_str),
             ),
         )
-        log.info("Saved to %s", self._serialization_dir)
+        log.info("Saved general checkpoints to %s", self._serialization_dir)
+
+        # For each of the tasks, we update the best task-speciic checkpoints
+        # only for target training.
+        if phase == "target_train":
+            self.save_task_specific_checkpoints(epoch, phase, task_states, model_state)
 
         if new_best_macro:
             self._unmark_previous_best(phase, epoch)
 
         if not self._keep_all_checkpoints:
             self._delete_old_checkpoints(phase, epoch)
 
+    def save_task_specific_checkpoints(self, epoch, phase, task_states, model_state):
+        """
+        Saves task specific checkpoints for all tasks where the validation performance improved. 
+
+        Parameters 
+        --------------------
+            - epoch: int
+            - task_states: dict, dictionary with task experiment parameters such as optimizer and scheduler. 
+            - model_state: experiment model
+        
+        Returns
+        --------------------
+        None
+        """
+        for metric_name, metric_info in self._metric_infos.items():
+            # if the best checpoint is thsi epoch
+            if metric_info["best"][0] == epoch and metric_info["task_name"] is not None:
+                task_name = metric_info["task_name"]
+                torch.save(
+                    metric_info,
+                    os.path.join(
+                        self._serialization_dir,
+                        task_name + "/metric_state_{}_epoch_{}.best.th".format(phase, epoch),
```

Comment:
We weren't previously saving metric/task state for target tasks, right? It's reasonable to fix that, but if we fix it here, make sure we can actually use the stuff we save.

Reply:
We were previously saving metric/task state for target tasks.

Reply:
Ah, yeah. Do we ever use them, though?